File: app/src/main/java/jp/hazuki/yuzubrowser/bookmark/view/AddBookmarkDialog.java
package jp.hazuki.yuzubrowser.bookmark.view;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.text.Editable;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import jp.hazuki.yuzubrowser.R;
import jp.hazuki.yuzubrowser.bookmark.BookmarkFolder;
import jp.hazuki.yuzubrowser.bookmark.BookmarkItem;
import jp.hazuki.yuzubrowser.bookmark.BookmarkManager;
import jp.hazuki.yuzubrowser.bookmark.view.BookmarkFoldersDialog.OnFolderSelectedListener;
import jp.hazuki.yuzubrowser.utils.view.SpinnerButton;

public abstract class AddBookmarkDialog<S extends BookmarkItem, T> implements OnFolderSelectedListener {
    protected final Context mContext;
    protected final AlertDialog mDialog;
    protected final EditText titleEditText;
    protected final EditText urlEditText;
    protected final SpinnerButton folderButton;
    protected DialogInterface.OnClickListener mOnClickListener;
    protected final S mItem;
    protected BookmarkManager mManager;
    protected BookmarkFolder mParent;

    public AddBookmarkDialog(final Context context, BookmarkManager manager, S item, String title, T url) {
        mContext = context;
        mItem = item;
        mManager = manager;

        if (mManager == null)
            mManager = new BookmarkManager(context);

        View view = inflateView();
        titleEditText = (EditText) view.findViewById(R.id.titleEditText);
        urlEditText = (EditText) view.findViewById(R.id.urlEditText);
        folderButton = (SpinnerButton) view.findViewById(R.id.folderButton);

        initView(view, title, url);

        //titleEditText.setText((title == null)?url:title);
        //urlEditText.setText(url);

        mDialog = new AlertDialog.Builder(context)
                .setTitle(R.string.add_bookmark)
                .setView(view)
                .setPositiveButton(android.R.string.ok, null)
                .setNegativeButton(android.R.string.cancel, null)
                .create();
    }

    protected Context getContext() {
        return mContext;
    }

    protected View inflateView() {
        return LayoutInflater.from(mContext).inflate(R.layout.add_bookmark_dialog, null);
    }

    protected void initView(View view, String title, T url) {
        if (mItem == null) {
            final BookmarkFolder root = mManager.getRoot();
            mParent = root;
            folderButton.setText(root.title);
            folderButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    new BookmarkFoldersDialog(mContext, mManager)
                            .setTitle(R.string.folder)
                            .setCurrentFolder(root)
                            .setOnFolderSelectedListener(AddBookmarkDialog.this)
                            .show();
                }
            });
        } else {
            folderButton.setVisibility(View.GONE);
        }
    }

    public void show() {
        mDialog.show();

        mDialog.getButton(DialogInterface.BUTTON_POSITIVE).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Editable title = titleEditText.getText();
                if (TextUtils.isEmpty(title)) {
                    Toast.makeText(mDialog.getContext(), R.string.title_empty_mes, Toast.LENGTH_SHORT).show();
                    return;
                }
                Editable url = urlEditText.getText();
                if (TextUtils.isEmpty(url)) {
                    Toast.makeText(mDialog.getContext(), R.string.url_empty_mes, Toast.LENGTH_SHORT).show();
                    return;
                }

                S item = makeItem(mItem, title.toString(), url.toString());
                if (item != null)
                    mParent.add(item);

                if (mManager.write()) {
                    Toast.makeText(mDialog.getContext(), R.string.succeed, Toast.LENGTH_SHORT).show();
                    if (mOnClickListener != null)
                        mOnClickListener.onClick(mDialog, DialogInterface.BUTTON_POSITIVE);
                    mDialog.dismiss();
                } else {
                    Toast.makeText(mDialog.getContext(), R.string.failed, Toast.LENGTH_LONG).show();
                }
            }
        });
    }

    protected abstract S makeItem(S item, String title, String url);

    public AddBookmarkDialog<S, T> setOnClickListener(DialogInterface.OnClickListener l) {
        mOnClickListener = l;
        return this;
    }

    @Override
    public boolean onFolderSelected(DialogInterface dialog, BookmarkFolder folder) {
        folderButton.setText(folder.title);
        mParent = folder;
        return false;
    }
}

File: module/bookmark/src/main/java/jp/hazuki/yuzubrowser/bookmark/view/AddBookmarkDialog.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.bookmark.view

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.widget.CheckBox
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import jp.hazuki.bookmark.R
import jp.hazuki.yuzubrowser.bookmark.item.BookmarkFolder
import jp.hazuki.yuzubrowser.bookmark.item.BookmarkItem
import jp.hazuki.yuzubrowser.bookmark.repository.BookmarkManager
import jp.hazuki.yuzubrowser.ui.BROADCAST_ACTION_NOTIFY_CHANGE_WEB_STATE
import jp.hazuki.yuzubrowser.ui.extensions.toPunyCodeUrl
import jp.hazuki.yuzubrowser.ui.settings.AppPrefs
import jp.hazuki.yuzubrowser.ui.widget.SpinnerButton

abstract class AddBookmarkDialog<S : BookmarkItem, T>(
    protected val context: Context,
    manager: BookmarkManager?,
    protected val mItem: S?,
    title: String?,
    url: T
) : BookmarkFoldersDialog.OnFolderSelectedListener {
    protected val mDialog: AlertDialog
    protected val titleEditText: EditText
    protected val urlEditText: EditText
    protected val folderTextView: TextView
    protected val folderButton: SpinnerButton
    protected val addToTopCheckBox: CheckBox
    protected var mOnClickListener: DialogInterface.OnClickListener? = null
    protected val mManager: BookmarkManager = manager ?: BookmarkManager.getInstance(context)
    protected lateinit var mParent: BookmarkFolder

    init {
        val view = inflateView()
        titleEditText = view.findViewById(R.id.titleEditText)
        urlEditText = view.findViewById(R.id.urlEditText)
        folderTextView = view.findViewById(R.id.folderTextView)
        folderButton = view.findViewById(R.id.folderButton)
        addToTopCheckBox = view.findViewById(R.id.addToTopCheckBox)

        initView(view, title, url)

        mDialog = AlertDialog.Builder(context)
                .setTitle(if (mItem == null) R.string.add_bookmark else R.string.edit_bookmark)
                .setView(view)
                .setPositiveButton(android.R.string.ok, null)
                .setNegativeButton(android.R.string.cancel, null)
                .create()
    }

    protected fun inflateView(): View {
        return LayoutInflater.from(context).inflate(R.layout.add_bookmark_dialog, null)
    }

    protected open fun initView(view: View, title: String?, url: T) {
        if (mItem == null) {
            val root = getRootPosition()
            mParent = root
            folderButton.text = root.title
            folderButton.setOnClickListener { v ->
                BookmarkFoldersDialog(context, mManager)
                        .setTitle(R.string.folder)
                        .setCurrentFolder(root)
                        .setOnFolderSelectedListener(this@AddBookmarkDialog)
                        .show()
            }
        } else {
            folderTextView.visibility = View.GONE
            folderButton.visibility = View.GONE
            addToTopCheckBox.visibility = View.GONE
        }
    }

    private fun getRootPosition(): BookmarkFolder {
        if (AppPrefs.saveBookmarkFolder.get()) {
            val id = AppPrefs.saveBookmarkFolderId.get()
            val item = mManager[id]
            if (item is BookmarkFolder) {
                return item
            }
        }
        return mManager.root
    }

    fun show() {
        mDialog.show()

        mDialog.getButton(DialogInterface.BUTTON_POSITIVE).setOnClickListener { v ->
            val title = titleEditText.text
            if (TextUtils.isEmpty(title)) {
                Toast.makeText(mDialog.context, R.string.title_empty_mes, Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }
            val url = urlEditText.text
            if (TextUtils.isEmpty(url)) {
                Toast.makeText(mDialog.context, R.string.url_empty_mes, Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            val item = makeItem(mItem, title.toString(), url.toString().toPunyCodeUrl())
            if (item != null) {
                if (addToTopCheckBox.isChecked)
                    mManager.addFirst(mParent, item)
                else
                    mManager.add(mParent, item)
            }
            if (mItem != null && addToTopCheckBox.isChecked) {
                mManager.moveToFirst(mParent, mItem)
            }

            if (mManager.save()) {
                Toast.makeText(mDialog.context, R.string.succeed, Toast.LENGTH_SHORT).show()
                mOnClickListener?.onClick(mDialog, DialogInterface.BUTTON_POSITIVE)
                LocalBroadcastManager.getInstance(context)
                    .sendBroadcast(Intent(BROADCAST_ACTION_NOTIFY_CHANGE_WEB_STATE))
                mDialog.dismiss()
            } else {
                Toast.makeText(mDialog.context, R.string.failed, Toast.LENGTH_LONG).show()
            }
        }
    }

    protected abstract fun makeItem(item: S?, title: String, url: String): S?

    fun setOnClickListener(l: DialogInterface.OnClickListener): AddBookmarkDialog<S, T> {
        mOnClickListener = l
        return this
    }

    inline fun setOnClickListener(crossinline listener: (DialogInterface, Int) -> Unit): AddBookmarkDialog<S, T> {
        setOnClickListener(DialogInterface.OnClickListener { dialog, which -> listener(dialog, which) })
        return this
    }

    override fun onFolderSelected(dialog: DialogInterface, folder: BookmarkFolder): Boolean {
        folderButton.text = folder.title
        mParent = folder
        return false
    }
}
